SLHCUpgradeSimulations/L1TrackTrigger/plugins/L1TkMuonFromExtendedProducer.cc

```diff
@@ -91,35 +140,145 @@ class L1TkMuonFromExtendedProducer : public edm::EDProducer {
+
+  // Muons filling
 
-   correctGMTPropForTkZ_ = iConfig.getParameter<bool>("correctGMTPropForTkZ");
+  int imu = -1;
+  int MuonNum = -1;
+  //  int TrackNum = -1;
+  for (auto l1mu : l1mus){
+    imu++;
 
-   use5ParameterFit_     = iConfig.getParameter<bool>("use5ParameterFit");
-   produces<L1TkMuonParticleCollection>();
+    if( imu<MuonMax ) {
+      MuonNum = imu + 1;
+      MuCn[imu].eta = l1mu.eta();
+      MuCn[imu].eta = l1mu.eta();
+      MuCn[imu].phi = l1mu.phi();
+      MuCn[imu].pt = l1mu.pt();
+      MuCn[imu].bx = l1mu.bx();
+      MuCn[imu].quality = l1mu.quality();
+      MuCn[imu].sigmaEta = l1mu.sigmaEta();
+      MuCn[imu].sigmaPhi = l1mu.sigmaPhi();
+      MuCn[imu].feta = fabs( l1mu.eta());
+  
+      const auto& gmtCand = l1mu.gmtMuonCand();
+      const auto& dtCand  = l1mu.dtCand();
+      const auto& cscCand = l1mu.cscCand();
+      const auto& rpcCand = l1mu.rpcCand();
+  
+      MuCn[imu].gmtCand = gmtCand.empty();
+      MuCn[imu].dtCand = dtCand.empty();
+      MuCn[imu].cscCand = cscCand.empty();
+      MuCn[imu].rpcCand = rpcCand.empty();
+  
+      LogDebug("L1TkMuonFromExtendedProducer") << " imu= " << imu
+        << " eta= " << MuCn[imu].eta 
+        << " phi= " << MuCn[imu].phi 
+        << " bx= " << MuCn[imu].bx
+        << " quality= " << MuCn[imu].quality
+        << " sigmaEta= " << MuCn[imu].sigmaEta
+        << " sigmaPhi= " << MuCn[imu].sigmaPhi 
+        << " gmtCand " << MuCn[imu].gmtCand 
+        << " dtCand " << MuCn[imu].dtCand 
+        << " cscCand " << MuCn[imu].cscCand 
+        << " rpcCand " << MuCn[imu].rpcCand;
+    } 
+    else { 
+      LogWarning("L1TkMuonFromExtendedProducer") << " Maximum Muon number too small " << imu;
+    };
+  }
+  LogDebug("L1TkMuonFromExtendedProducer") << " ::keepMuons MuonNum " << MuonNum;
+  return MuonNum;
 }
 
-L1TkMuonFromExtendedProducer::~L1TkMuonFromExtendedProducer() {
+// ------------ Tracks converter  ------------
+int
+L1TkMuonFromExtendedProducer::keepTracks(edm::Event& iEvent, L1Tracks* L1Tk)
+{
+  using namespace edm;
+  using namespace std;
+
+  edm::Handle<L1TkTrackCollectionType> l1tksH;
+  iEvent.getByLabel(L1TrackInputTag_, l1tksH);
+  const L1TkTrackCollectionType& l1tks = *l1tksH.product();
+
+  int TrackNum = -1;
+  int il1tk = -1;
+  for (auto l1tk : l1tks ){
+    il1tk++;
+  
+    if( il1tk<TrackMax ) {
+
+      TrackNum = il1tk + 1;
+
+      unsigned int nPars = 4;
+      if (use5ParameterFit_) nPars = 5;
+      L1Tk[il1tk].nPars = nPars;
+
+      L1Tk[il1tk].pt = l1tk.getMomentum(nPars).perp();
+      L1Tk[il1tk].z  = l1tk.getPOCA(nPars).z();
+      L1Tk[il1tk].chi2 = l1tk.getChi2(nPars);
+      L1Tk[il1tk].nstubs = l1tk.getStubRefs().size();
+      L1Tk[il1tk].eta = l1tk.getMomentum(nPars).eta();
+      L1Tk[il1tk].phi = l1tk.getMomentum(nPars).phi();
+
+      L1Tk[il1tk].idx = il1tk;
+
+      PropState pstate = propagateToGMT(l1tk);
+      if (!pstate.valid) continue;
+
+      L1Tk[il1tk].matchPt = pstate.pt; 
```

Comment:
better don't skip, just fill the bad default values (should come for a failed case).